README.md
# AES Haskell
An implementation of the Advanced Encryption Standard in Haskell.

**This implementation is not cryptographically secure.**

## Usage
If you don't have Stack installed, find out how to get it [here](https://docs.haskellstack.org/en/stable/README/).

Run the program with `stack run --`, followed by your options.

```
Usage: aes-exe (-k|--key KEY) ((-f|--file FILENAME) | (-r|--raw DATA)) 
               [-d|--decrypt] [-s]
  128, 192 and 256-bit ECB AES implementation.

Available options:
  -h,--help                Show this help text
  -k,--key KEY             Key as a hexadecimal string
  -f,--file FILENAME       Input file path
  -r,--raw DATA            Input as hexadecimal string
  -d,--decrypt             Decrypt
  -s                       Output as text
```

### Example

To encrypt the hexadecimal `3243f6a8885a308d313198a2e0370734` with key `2b7e151628aed2a6abf7158809cf4f3c` (taken from Appendix B of the standard), you would run
```
stack run -- -k "2b7e151628aed2a6abf7158809cf4f3c" -r "3243f6a8885a308d313198a2e0370734"
```
which should give `3925841d02dc09fbdc118597196a0b32`. To decrypt this, run
```
stack run -- -k "2b7e151628aed2a6abf7158809cf4f3c" -r "3925841d02dc09fbdc118597196a0b32" -d
```
which should give back the original.


ChangeLog.md
# Changelog for aes

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 1.0 - 2022-09-01
Initial complete implementation with CLI.


test/Spec.hs
import Test.Tasty ( testGroup, defaultMain, TestTree )
import Test.Tasty.HUnit ( assertEqual, testCase )
import Types (Byte, KeySchedule, Key, State)
import Cipher (keyExpansion, cipher, invCipher)

-- Appendix A tests
a1InputKey :: Key
a1InputKey = [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6, 0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c]
a1Output :: KeySchedule
a1Output = [[[0x2b, 0x28, 0xab, 0x09], [0x7e, 0xae, 0xf7, 0xcf], [0x15, 0xd2, 0x15, 0x4f], [0x16, 0xa6, 0x88, 0x3c]], [[0xa0, 0x88, 0x23, 0x2a], [0xfa, 0x54, 0xa3, 0x6c], [0xfe, 0x2c, 0x39, 0x76], [0x17, 0xb1, 0x39, 0x05]], [[0xf2, 0x7a, 0x59, 0x73], [0xc2, 0x96, 0x35, 0x59], [0x95, 0xb9, 0x80, 0xf6], [0xf2, 0x43, 0x7a, 0x7f]], [[0x3d, 0x47, 0x1e, 0x6d], [0x80, 0x16, 0x23, 0x7a], [0x47, 0xfe, 0x7e, 0x88], [0x7d, 0x3e, 0x44, 0x3b]], [[0xef, 0xa8, 0xb6, 0xdb], [0x44, 0x52, 0x71, 0x0b], [0xa5, 0x5b, 0x25, 0xad], [0x41, 0x7f, 0x3b, 0x00]], [[0xd4, 0x7c, 0xca, 0x11], [0xd1, 0x83, 0xf2, 0xf9], [0xc6, 0x9d, 0xb8, 0x15], [0xf8, 0x87, 0xbc, 0xbc]], [[0x6d, 0x11, 0xdb, 0xca], [0x88, 0x0b, 0xf9, 0x00], [0xa3, 0x3e, 0x86, 0x93], [0x7a, 0xfd, 0x41, 0xfd]], [[0x4e, 0x5f, 0x84, 0x4e], [0x54, 0x5f, 0xa6, 0xa6], [0xf7, 0xc9, 0x4f, 0xdc], [0x0e, 0xf3, 0xb2, 0x4f]], [[0xea, 0xb5, 0x31, 0x7f], [0xd2, 0x8d, 0x2b, 0x8d], [0x73, 0xba, 0xf5, 0x29], [0x21, 0xd2, 0x60, 0x2f]], [[0xac, 0x19, 0x28, 0x57], [0x77, 0xfa, 0xd1, 0x5c], [0x66, 0xdc, 0x29, 0x00], [0xf3, 0x21, 0x41, 0x6e]], [[0xd0, 0xc9, 0xe1, 0xb6], [0x14, 0xee, 0x3f, 0x63], [0xf9, 0x25, 0x0c, 0x0c], [0xa8, 0x89, 0xc8, 0xa6]]]
a2InputKey :: Key
a2InputKey = [0x8e, 0x73, 0xb0, 0xf7, 0xda, 0x0e, 0x64, 0x52, 0xc8, 0x10, 0xf3, 0x2b, 0x80, 0x90, 0x79, 0xe5, 0x62, 0xf8, 0xea, 0xd2, 0x52, 0x2c, 0x6b, 0x7b]
a2Output :: KeySchedule
a2Output = [[[0x8e, 0xda, 0xc8, 0x80], [0x73, 0x0e, 0x10, 0x90], [0xb0, 0x64, 0xf3, 0x79], [0xf7, 0x52, 0x2b, 0xe5]], [[0x62, 0x52, 0xfe, 0x24], [0xf8, 0x2c, 0x0c, 0x02], [0xea, 0x6b, 0x91, 0xf5], [0xd2, 0x7b, 0xf7, 0xa5]], [[0xec, 0x6c, 0x0e, 0x5c], [0x12, 0x82, 0x7a, 0x56], [0x06, 0x7f, 0x95, 0xfe], [0x8e, 0x6b, 0xb9, 0xc2]], [[0x4d, 0x69, 0x85, 0xe9], [0xb7, 0xb5, 0xa7, 0x25], [0xb4, 0x41, 0x47, 0x38], [0xbd, 0x18, 0x96, 0xfd]], [[0xe7, 0xbb, 0x48, 0x21], [0x5f, 0x09, 0x5a, 0xef], [0xad, 0x53, 0xf0, 0xb1], [0x44, 0x86, 0x57, 0x4f]], [[0xa4, 0x4d, 0xaa, 0x11], [0x48, 0x6d, 0x32, 0x3b], [0xf6, 0xce, 0x63, 0x30], [0xd9, 0x24, 0x60, 0xe6]], [[0xa2, 0x83, 0x27, 0x6a], [0x5e, 0xb1, 0xf9, 0x94], [0x7e, 0xcf, 0x39, 0xf7], [0xd5, 0x9a, 0x43, 0x67]], [[0xc0, 0xd1, 0xec, 0x6f], [0xa6, 0x9d, 0x17, 0xa6], [0x94, 0xa4, 0x86, 0x49], [0x07, 0xe1, 0xeb, 0x71]], [[0x48, 0x22, 0xe2, 0x33], [0x5f, 0xcb, 0x6d, 0xf0], [0x70, 0x87, 0x13, 0xb7], [0x32, 0x55, 0x52, 0xb3]], [[0x40, 0x2f, 0x67, 0x45], [0xbe, 0x18, 0x47, 0x8c], [0xeb, 0xa2, 0xd2, 0x55], [0x28, 0x59, 0x6b, 0x3e]], [[0xa7, 0x94, 0x82, 0xad], [0xe1, 0x11, 0x1f, 0x07], [0x46, 0xf1, 0x75, 0xd7], [0x6c, 0xdf, 0x0a, 0x53]], [[0xca, 0x8f, 0x28, 0xbc], [0x40, 0xcc, 0x2d, 0x3c], [0x05, 0x50, 0x16, 0xe7], [0x38, 0x06, 0x6a, 0xb5]], [[0xe9, 0x44, 0x8e, 0x01], [0x8b, 0x8c, 0xcc, 0x00], [0xa0, 0x77, 0x72, 0x22], [0x6f, 0x3c, 0x04, 0x02]]]
a3InputKey :: Key
a3InputKey = [0x60, 0x3d, 0xeb, 0x10, 0x15, 0xca, 0x71, 0xbe, 0x2b, 0x73, 0xae, 0xf0, 0x85, 0x7d, 0x77, 0x81, 0x1f, 0x35, 0x2c, 0x07, 0x3b, 0x61, 0x08, 0xd7, 0x2d, 0x98, 0x10, 0xa3, 0x09, 0x14, 0xdf, 0xf4]
a3Output :: KeySchedule
a3Output = [[[0x60, 0x15, 0x2b, 0x85], [0x3d, 0xca, 0x73, 0x7d], [0xeb, 0x71, 0xae, 0x77], [0x10, 0xbe, 0xf0, 0x81]], [[0x1f, 0x3b, 0x2d, 0x09], [0x35, 0x61, 0x98, 0x14], [0x2c, 0x08, 0x10, 0xdf], [0x07, 0xd7, 0xa3, 0xf4]], [[0x9b, 0x8e, 0xa5, 0x20], [0xa3, 0x69, 0x1a, 0x67], [0x54, 0x25, 0x8b, 0xfc], [0x11, 0xaf, 0x5f, 0xde]], [[0xa8, 0x93, 0xbe, 0xb7], [0xb0, 0xd1, 0x49, 0x5d], [0x9c, 0x94, 0x84, 0x5b], [0x1a, 0xcd, 0x6e, 0x9a]], [[0xd5, 0x5b, 0xfe, 0xde], [0x9a, 0xf3, 0xe9, 0x8e], [0xec, 0xc9, 0x42, 0xbe], [0xb8, 0x17, 0x48, 0x96]], [[0xb5, 0x26, 0x98, 0x2f], [0xa9, 0x78, 0x31, 0x6c], [0x32, 0xa6, 0x22, 0x79], [0x8a, 0x47, 0x29, 0xb3]], [[0x81, 0xda, 0x24, 0xfa], [0x2c, 0xdf, 0x36, 0xb8], [0x81, 0x48, 0x0a, 0xb4], [0xad, 0xba, 0xf2, 0x64]], [[0x98, 0xbe, 0x26, 0x09], [0xc5, 0xbd, 0x8c, 0xe0], [0xbf, 0x19, 0x3b, 0x42], [0xc9, 0x8e, 0xa7, 0x14]], [[0x68, 0xb2, 0x96, 0x6c], [0x00, 0xdf, 0xe9, 0x51], [0x7b, 0x33, 0x39, 0x8d], [0xac, 0x16, 0xe4, 0x80]], [[0xc8, 0x76, 0x50, 0x59], [0x14, 0xa9, 0x25, 0xc5], [0xe2, 0xfb, 0xc0, 0x82], [0x04, 0x8a, 0x2d, 0x39]], [[0xde, 0x6c, 0xfa, 0x96], [0x13, 0xcc, 0x25, 0x74], [0x69, 0x5a, 0x63, 0xee], [0x67, 0x71, 0x95, 0x15]], [[0x58, 0x2e, 0x7e, 0x27], [0x86, 0x2f, 0x0a, 0xcf], [0xca, 0x31, 0xf1, 0x73], [0x5d, 0xd7, 0xfa, 0xc3]], [[0x74, 0x18, 0xe2, 0x74], [0x9c, 0x50, 0x75, 0x01], [0x47, 0x1d, 0x7e, 0x90], [0xab, 0xda, 0x4f, 0x5a]], [[0xca, 0xe4, 0x9a, 0xbd], [0xfa, 0xd5, 0xdf, 0x10], [0xaa, 0x9b, 0x6a, 0x19], [0xe3, 0x34, 0xce, 0x0d]], [[0xfe, 0xe6, 0x04, 0x70], [0x48, 0x18, 0x6d, 0x6c], [0x90, 0x8d, 0xf3, 0x63], [0xd1, 0x0b, 0x44, 0x1e]]]

-- Appendix B test
bInput :: State
bInput = [[0x32, 0x88, 0x31, 0xe0], [0x43, 0x5a, 0x31, 0x37], [0xf6, 0x30, 0x98, 0x07], [0xa8, 0x8d, 0xa2, 0x34]]
bOutput :: State
bOutput = [[0x39, 0x02, 0xdc, 0x19], [0x25, 0xdc, 0x11, 0x6a], [0x84, 0x09, 0x85, 0x0b], [0x1d, 0xfb, 0x97, 0x32]]

-- Appendix C tests
cInput :: State
cInput = [[0x00, 0x44, 0x88, 0xcc],[0x11, 0x55, 0x99, 0xdd],[0x22, 0x66, 0xaa, 0xee],[0x33, 0x77, 0xbb, 0xff]]
c1InputKey :: Key
c1InputKey = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f]
c1Output :: State
c1Output = [[0x69, 0x6a, 0xd8, 0x70], [0xc4, 0x7b, 0xcd, 0xb4], [0xe0, 0x04, 0xb7, 0xc5], [0xd8, 0x30, 0x80, 0x5a]]
c2InputKey :: Key
c2InputKey = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17]
c2Output :: State
c2Output = [[0xdd, 0x86, 0x6e, 0xec], [0xa9, 0x4c, 0xaf, 0x0d], [0x7c, 0xdf, 0x70, 0x71], [0xa4, 0xe0, 0xa0, 0x91]]
c3InputKey :: Key
c3InputKey = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f]
c3Output :: State
c3Output = [[0x8e, 0x51, 0xea, 0x4b], [0xa2, 0x67, 0xfc, 0x49], [0xb7, 0x45, 0x49, 0x60], [0xca, 0xbf, 0x90, 0x89]]


main :: IO ()
main = defaultMain tests

tests :: TestTree
tests = testGroup "Tests" [appendixA, appendixB, appendixC]

appendixA :: TestTree
appendixA = testGroup "Appendix A"
    [
        testCase "Appendix A.1 - Expansion of 128-bit key" $
            assertEqual "Key schedules do not match" a1Output (keyExpansion a1InputKey),
        testCase "Appendix A.2 - Expansion of 192-bit key" $
            assertEqual "Key schedules do not match" a2Output (keyExpansion a2InputKey),
        testCase "Appendix A.3 - Expansion of 256-bit key" $
            assertEqual "Key schedules do not match" a3Output (keyExpansion a3InputKey)
    ]

appendixB :: TestTree
appendixB = testGroup "Appendix B"
    [
        testCase "Appendix B - 128-bit cipher" $
            assertEqual "States do not match" bOutput (cipher bInput a1InputKey)
    ]

appendixC :: TestTree
appendixC = testGroup "Appendix C" 
    [
        testCase "Appendix C.1 - 128-bit cipher" $
            assertEqual "States do not match" c1Output (cipher cInput c1InputKey),
        testCase "Appendix C.1 - 128-bit inverse cipher" $
            assertEqual "States do not match" cInput (invCipher c1Output c1InputKey),
        testCase "Appendix C.2 - 192-bit cipher" $
            assertEqual "States do not match" c2Output (cipher cInput c2InputKey),
        testCase "Appendix C.2 - 192-bit inverse cipher" $
            assertEqual "States do not match" cInput (invCipher c2Output c2InputKey),
        testCase "Appendix C.3 - 256-bit cipher" $
            assertEqual "States do not match" c3Output (cipher cInput c3InputKey),
        testCase "Appendix C.3 - 256-bit inverse cipher" $
            assertEqual "States do not match" cInput (invCipher c3Output c3InputKey)
    ]


app/Main.hs
module Main where

import Options.Applicative
    ( Alternative((<|>)),
      help,
      info,
      long,
      metavar,
      progDesc,
      short,
      strOption,
      switch,
      execParser,
      helper,
      Parser )
import Data.Semigroup ((<>))
import Cipher (cipher, invCipher)
import qualified Data.ByteString as BS (readFile, unpack)
import Util (fromHex, inputChunker, outputDeChunker, toHex, toString)

data Options = Options {
    key        :: String,
    input      :: Input,
    decrypt    :: Bool,
    asString   :: Bool
}

-- Input to the program can either be a file or raw input
data Input = FileInput FilePath | RawInput String deriving Show

fileInput :: Parser Input
fileInput = FileInput <$> strOption (
    long "file" <>
    short 'f' <>
    metavar "FILENAME" <>
    help "Input file path")

rawInput :: Parser Input
rawInput = RawInput <$> strOption (
    long "raw" <>
    short 'r' <>
    metavar "DATA" <>
    help "Input as hexadecimal string")

inputParser :: Parser Input
inputParser = fileInput <|> rawInput

optionParser :: Parser Options
optionParser = Options <$>
    strOption (
        long "key" <>
        short 'k' <>
        metavar "KEY" <>
        help "Key as a hexadecimal string"
    ) <*>
    inputParser <*>
    switch (
        long "decrypt" <>
        short 'd' <>
        help "Decrypt"
    ) <*>
    switch (
        short 's' <>
        help "Output as text"
    )

main :: IO ()
main = do
    -- Parse args
    args <- execParser $ info (helper <*> optionParser) (progDesc "128, 192 and 256-bit ECB AES implementation.")
    -- Either get the raw text or read the file
    states <- case input args of
            -- Convert raw string to State
            RawInput r -> pure $ inputChunker $ fromHex r
            -- Read the given file as a bytestring, unpack and convert to State
            FileInput f -> do
                content <- BS.readFile f
                pure $ inputChunker $ BS.unpack content
    -- Get key as [Byte]
    let k = fromHex $ key args
    -- Get result
    let result = if decrypt args
        then map (`invCipher` k) states
        else map (`cipher` k) states
    -- Determine whether to output as hex or text
    let outputFormatter = if asString args then toString else toHex
    -- Flatten output, format and print
    putStrLn $ outputFormatter $ outputDeChunker result

src/Util.hs
module Util where
import Numeric (showHex, readHex)
import Data.List (transpose)
import Data.List.Split (chunksOf)
import Data.Char (chr)

import Types ( Byte, State ) 


-- Chunk a list into a 2d list with items of size 4
chunk :: [a] -> [[a]]
chunk y =  map (take 4 . flip drop y . (*4)) [0..(length y `div` 4)-1]

-- Convert a list of bytes to a list of states, padding with 0.
inputChunker :: [Byte] -> [State]
inputChunker x = map toState $ chunksOf 16 (x ++ getPadding x) where
    getPadding :: [Byte] -> [Byte]
    getPadding x = replicate n 0 where
        a = length x `mod` 16
        n = if a == 0 then 0 else 16 - a 

outputDeChunker :: [State] -> [Byte]
outputDeChunker = concatMap fromState

-- Convert hex string to state
hexToState :: String -> State
hexToState = toState . fromHex

-- Convert state to hex
stateToHex :: State -> String
stateToHex = toHex . fromState

-- Convert hex string WITHOUT SPACES to list of bytes
fromHex :: String -> [Byte]
fromHex (x:y:xs) = fst (head $ readHex [x, y]) : fromHex xs
fromHex _ =  []

-- Converts a list of bytes to a state
toState :: [Byte] -> State
toState = transpose . chunk

-- Convert a state to a list of bytes
fromState :: State -> [Byte]
fromState = concat . transpose

-- Convert list of bytes to hex string
toHex :: [Byte] -> String
toHex = concatMap (\x -> (if x < 0x10 then "0" else "") ++ showHex x "")

-- Convert list of bytes to actual string
toString :: [Byte] -> String
toString = map (chr . fromEnum)

-- Convert a state to hex strings
stateHex :: State -> [[String]]
stateHex = (map.map) (`showHex` "")

src/Types.hs
module Types where

import Data.Word ( Word8 )

-- Byte is just a synonym for Word8. Makes code more readable with the spec
type Byte = Word8

-- In the specification, a word refers to 4 bytes or 32 bits
type Word = [Byte]

-- The state is a 4x4 array of bytes, representing 128 bits of data
type State = [[Byte]]

-- The encryption key is stored as an array of bytes
type Key = [Byte]

-- The KeySchedule will be an (Nr+1) * 4 * 4 array of bytes
-- Each 4*4 sub-list is a Round Key, and there are Nr + 1 round keys
-- A round key is the same size as the state, which it is then XOR'd with
{-
For example:
    [
        [[a,b,c,d],[e,f,g,h],[i,j,k,l],[m,n,o,p]], -- Round key 1
        ...,
        [[a,b,c,d],[e,f,g,h],[i,j,k,l],[m,n,o,p]]  -- Round key Nr + 1
    ]
-}
type RoundKey = [[Byte]]
type KeySchedule = [RoundKey]

src/Cipher.hs
module Cipher (cipher, invCipher, keyExpansion) where
import Data.Word (Word8)
import Data.Bits (Bits(xor, shiftR, (.&.), shiftL))
import Data.List (transpose)
import Control.Arrow ((>>>))
import Prelude hiding (Word)

import Types (Key, KeySchedule, RoundKey, State, Word, Byte)
import Util (chunk)

-- Nk is the number of 32-bit words in the key
-- 32 bits = 4*8 bits, hence the div 4
nk :: Key -> Int
nk = (`div` 4) . length
-- Nb is the number of columns in the state, which is always 4
nb :: Int
nb = 4
-- Nr is the number of rounds, it is the key length + 6
nr :: Key -> Int
nr = (+6) . nk

-- The substitution box
sbox :: [Byte]
sbox = [0x63, 0x7c, 0x77, 0x7b, 0xf2, 0x6b, 0x6f, 0xc5, 0x30, 0x01, 0x67, 0x2b, 0xfe, 0xd7, 0xab, 0x76, 0xca, 0x82, 0xc9, 0x7d, 0xfa, 0x59, 0x47, 0xf0, 0xad, 0xd4, 0xa2, 0xaf, 0x9c, 0xa4, 0x72, 0xc0, 0xb7, 0xfd, 0x93, 0x26, 0x36, 0x3f, 0xf7, 0xcc, 0x34, 0xa5, 0xe5, 0xf1, 0x71, 0xd8, 0x31, 0x15, 0x04, 0xc7, 0x23, 0xc3, 0x18, 0x96, 0x05, 0x9a, 0x07, 0x12, 0x80, 0xe2, 0xeb, 0x27, 0xb2, 0x75, 0x09, 0x83, 0x2c, 0x1a, 0x1b, 0x6e, 0x5a, 0xa0, 0x52, 0x3b, 0xd6, 0xb3, 0x29, 0xe3, 0x2f, 0x84, 0x53, 0xd1, 0x00, 0xed, 0x20, 0xfc, 0xb1, 0x5b, 0x6a, 0xcb, 0xbe, 0x39, 0x4a, 0x4c, 0x58, 0xcf, 0xd0, 0xef, 0xaa, 0xfb, 0x43, 0x4d, 0x33, 0x85, 0x45, 0xf9, 0x02, 0x7f, 0x50, 0x3c, 0x9f, 0xa8, 0x51, 0xa3, 0x40, 0x8f, 0x92, 0x9d, 0x38, 0xf5, 0xbc, 0xb6, 0xda, 0x21, 0x10, 0xff, 0xf3, 0xd2, 0xcd, 0x0c, 0x13, 0xec, 0x5f, 0x97, 0x44, 0x17, 0xc4, 0xa7, 0x7e, 0x3d, 0x64, 0x5d, 0x19, 0x73, 0x60, 0x81, 0x4f, 0xdc, 0x22, 0x2a, 0x90, 0x88, 0x46, 0xee, 0xb8, 0x14, 0xde, 0x5e, 0x0b, 0xdb, 0xe0, 0x32, 0x3a, 0x0a, 0x49, 0x06, 0x24, 0x5c, 0xc2, 0xd3, 0xac, 0x62, 0x91, 0x95, 0xe4, 0x79, 0xe7, 0xc8, 0x37, 0x6d, 0x8d, 0xd5, 0x4e, 0xa9, 0x6c, 0x56, 0xf4, 0xea, 0x65, 0x7a, 0xae, 0x08, 0xba, 0x78, 0x25, 0x2e, 0x1c, 0xa6, 0xb4, 0xc6, 0xe8, 0xdd, 0x74, 0x1f, 0x4b, 0xbd, 0x8b, 0x8a, 0x70, 0x3e, 0xb5, 0x66, 0x48, 0x03, 0xf6, 0x0e, 0x61, 0x35, 0x57, 0xb9, 0x86, 0xc1, 0x1d, 0x9e, 0xe1, 0xf8, 0x98, 0x11, 0x69, 0xd9, 0x8e, 0x94, 0x9b, 0x1e, 0x87, 0xe9, 0xce, 0x55, 0x28, 0xdf, 0x8c, 0xa1, 0x89, 0x0d, 0xbf, 0xe6, 0x42, 0x68, 0x41, 0x99, 0x2d, 0x0f, 0xb0, 0x54, 0xbb, 0x16]

-- The inverse substitution box
isbox :: [Byte]
isbox = [0x52, 0x09, 0x6a, 0xd5, 0x30, 0x36, 0xa5, 0x38, 0xbf, 0x40, 0xa3, 0x9e, 0x81, 0xf3, 0xd7, 0xfb, 0x7c, 0xe3, 0x39, 0x82, 0x9b, 0x2f, 0xff, 0x87, 0x34, 0x8e, 0x43, 0x44, 0xc4, 0xde, 0xe9, 0xcb, 0x54, 0x7b, 0x94, 0x32, 0xa6, 0xc2, 0x23, 0x3d, 0xee, 0x4c, 0x95, 0x0b, 0x42, 0xfa, 0xc3, 0x4e, 0x08, 0x2e, 0xa1, 0x66, 0x28, 0xd9, 0x24, 0xb2, 0x76, 0x5b, 0xa2, 0x49, 0x6d, 0x8b, 0xd1, 0x25, 0x72, 0xf8, 0xf6, 0x64, 0x86, 0x68, 0x98, 0x16, 0xd4, 0xa4, 0x5c, 0xcc, 0x5d, 0x65, 0xb6, 0x92, 0x6c, 0x70, 0x48, 0x50, 0xfd, 0xed, 0xb9, 0xda, 0x5e, 0x15, 0x46, 0x57, 0xa7, 0x8d, 0x9d, 0x84, 0x90, 0xd8, 0xab, 0x00, 0x8c, 0xbc, 0xd3, 0x0a, 0xf7, 0xe4, 0x58, 0x05, 0xb8, 0xb3, 0x45, 0x06, 0xd0, 0x2c, 0x1e, 0x8f, 0xca, 0x3f, 0x0f, 0x02, 0xc1, 0xaf, 0xbd, 0x03, 0x01, 0x13, 0x8a, 0x6b, 0x3a, 0x91, 0x11, 0x41, 0x4f, 0x67, 0xdc, 0xea, 0x97, 0xf2, 0xcf, 0xce, 0xf0, 0xb4, 0xe6, 0x73, 0x96, 0xac, 0x74, 0x22, 0xe7, 0xad, 0x35, 0x85, 0xe2, 0xf9, 0x37, 0xe8, 0x1c, 0x75, 0xdf, 0x6e, 0x47, 0xf1, 0x1a, 0x71, 0x1d, 0x29, 0xc5, 0x89, 0x6f, 0xb7, 0x62, 0x0e, 0xaa, 0x18, 0xbe, 0x1b, 0xfc, 0x56, 0x3e, 0x4b, 0xc6, 0xd2, 0x79, 0x20, 0x9a, 0xdb, 0xc0, 0xfe, 0x78, 0xcd, 0x5a, 0xf4, 0x1f, 0xdd, 0xa8, 0x33, 0x88, 0x07, 0xc7, 0x31, 0xb1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xec, 0x5f, 0x60, 0x51, 0x7f, 0xa9, 0x19, 0xb5, 0x4a, 0x0d, 0x2d, 0xe5, 0x7a, 0x9f, 0x93, 0xc9, 0x9c, 0xef, 0xa0, 0xe0, 0x3b, 0x4d, 0xae, 0x2a, 0xf5, 0xb0, 0xc8, 0xeb, 0xbb, 0x3c, 0x83, 0x53, 0x99, 0x61, 0x17, 0x2b, 0x04, 0x7e, 0xba, 0x77, 0xd6, 0x26, 0xe1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0c, 0x7d]

-- Key expansion generates Nr + 1 RoundKeys
keyExpansion :: Key -> KeySchedule
keyExpansion k = roundify $ expand (nk k) $ chunk k where
    -- SubWord is SubBytes but on a 1d list
    subWord :: Word -> Word
    subWord = map ((!!) sbox . fromIntegral)
    -- rotWord cyclically rotates a word left by 1
    rotWord :: Word -> Word
    rotWord = take 4 . drop 1 . cycle
    -- Round constant array (index 0 is never used)
    rcon :: Int -> Byte
    rcon = (!!) [0x00, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36]
    -- Convert a list of words to a key schedule
    roundify :: [Word] -> KeySchedule
    roundify [] = []
    roundify x  = transpose (take 4 x) : roundify (drop 4 x)
    -- Expand the key into a list of words
    expand :: Int -> [Word] -> [Word]
    expand n ws
        | n == nb * (nr k + 1) = ws
        | otherwise          = expand (n+1) (ws ++ [expand' n ws])
    -- expand' gives the value of the ith word of the key schedule
    expand' :: Int -> [Word] -> Word
    expand' i xs = zipWith xor (xs !! (i-nk k)) (temp i xs) where
        temp :: Int -> [Word] -> Word
        temp i ws
            | i `mod` nk k == 0           = xorFirst (subWord (rotWord temp')) (rcon (i `div` nk k))
            | nk k > 6 && i `mod` nk k == 4 = subWord temp'
            | otherwise                 = temp'
            where
                -- temp' is the initial value of temp
                temp' = ws !! (i-1)
                -- XORs the given byte with the first byte of the given word.
                xorFirst :: Word -> Byte -> Word
                xorFirst []     _ = []
                xorFirst (x:xs) v = x `xor` v : xs

cipher :: State -> Key -> State
cipher state key =
    if length state == 4 && length (head state) == 4 && length key `elem` [16, 24, 32]
    then roundFunction 1 initState
    else error "Invalid key or state size"
    where
    initState = addRoundKey (head keySchedule) state
    keySchedule = keyExpansion key
    roundFunction :: Int -> State -> State
    roundFunction n =
        -- The final round does not have MixColumns
        if n == nr key then subBytes sbox >>> shiftRows >>> addRoundKey (keySchedule !! n)
        else roundFunction (n+1) . (subBytes sbox >>> shiftRows >>> mixColumns >>> addRoundKey (keySchedule !! n))

invCipher :: State -> Key -> State
invCipher state key = 
    if length state == 4 && length (head state) == 4 && length key `elem` [16, 24, 32]
    then roundFunction (nr key - 1) initState
    else error "Invalid key or state size"
    where
    initState = addRoundKey (keySchedule !! nr key) state
    keySchedule = keyExpansion key
    roundFunction :: Int -> State -> State
    roundFunction n =
        -- The final round does not have InvMixColumns
        if n == 0 then invShiftRows >>> subBytes isbox >>> addRoundKey (keySchedule !! n)
        else roundFunction (n-1) . (invShiftRows >>> subBytes isbox >>> addRoundKey (keySchedule !! n) >>> invMixColumns)

-- Subsitute each byte using the given sbox
subBytes :: [Byte] -> State -> State
subBytes box = (map . map) ((!!) box . fromIntegral)

-- Cyclically shift each row, the first by 0, second by 1, third by 2, fourth by 3.
shiftRows :: State -> State
shiftRows st = zipWith (\row shift -> take nb $ drop shift $ cycle row) st [0..3]

invShiftRows :: State -> State
invShiftRows st = zipWith (\row shift -> take nb $ drop shift $ cycle row) st [0,3,2,1]

genericMixColumns :: [[Byte -> Byte]] -> State -> State
genericMixColumns ops = transpose . map mix . transpose where
    -- Every row of operations is applied to the column.
    -- The results of the operations are XOR'd to give a new column.
    mix :: [Byte] -> [Byte]
    mix col = map (foldr xor 0 . flip (zipWith ($)) col) ops

-- The column operations all follow the pattern
-- a b c d
-- d a b c
-- c d a b
-- b c d a
-- This function generates the 2d list of operations given its first row.
generateOperations :: [a] -> [[a]]
generateOperations x = map (\y -> take 4 $ drop y $ cycle x) [0,3,2,1]

mixColumns :: State -> State
mixColumns = genericMixColumns $ generateOperations [galoisMultiply 2, galoisMultiply 3, id, id]

invMixColumns :: State -> State
invMixColumns = genericMixColumns $ generateOperations $ map galoisMultiply [0x0e, 0x0b, 0x0d, 0x09]

-- Multiply two bytes in GF(256)
galoisMultiply :: Byte -> Byte -> Byte
galoisMultiply a b = result $ iterate (step1 >>> step2 >>> step3) (a, b, 0, 0) !! 8 where
    result (_, _, p, _) = p
    {-
        if((b & 1) == 1){
            p ^= a;
        }
    -}
    step1 (a, b, p, c) = (a, b, if (b .&. 1) == 1 then p `xor` a else p, c)
    {-
        b >>= 1;
        carry = a & 0x80;
        a <<= 1;
    -}
    step2 (a, b, p, c) = (shiftL a 1, shiftR b 1, p, a .&. 0x80)
    {-
        if(carry == 0x80){
            a ^= 0x1b;
        }
    -}
    step3 (a, b, p, c) = (if c == 0x80 then a `xor` 0x1b else a, b, p, c)

-- XORs the state with the round key
addRoundKey :: RoundKey -> State -> State
addRoundKey = zipWith (zipWith xor)
